simulator/infra/cache/cache_tag_array.h

```diff
@@ -1,82 +1,58 @@
 /**
- * cache_tag_array.h
- * Header for the cache tag array model.
- * MIPT-MIPS Assignment 5.
- * Ladin Oleg.
+ * CacheTagArray
+ * @author Oleg Ladin, Denis Los
  */
 
-/* Protection from multi-including. */
 #ifndef CACHE_TAG_ARRAY_H
 #define CACHE_TAG_ARRAY_H
 
-/* C++ libraries. */
-#include <list>
-#include <vector>
 
-/* Simulator modules. */
+#include <infra/cache/LRUTagCache.h>
 #include <infra/types.h>
 #include <infra/log.h>
 
-/* Replacement algorithm modules (LRU). */
-class LRUInfo
-{
-    /*
-     * "lru" contains sequences of using for each set. First number in set list
-     * is the least used.
-     */
-    std::vector< std::list< uint32>> lru = {};
-
-public:
-    LRUInfo( uint32 ways, uint32 sets);
-    void touch( uint32 set, uint32 way);
-    uint32 update( uint32 set);
-};
-
-class CacheTagArrayCheck : private Log
-{
-protected:
-    CacheTagArrayCheck( uint32 size_in_bytes,
-                        uint32 ways,
-                        uint32 line_size,
-                        uint32 addr_size_in_bits);
-public:
-        const uint32 size_in_bytes;
-        const uint32 ways;
-        const uint32 line_size;
-        const uint32 addr_size_in_bits;
+#include <vector>
+#include <map>
+#include <utility>
 
-};
 
-class CacheTagArray : public CacheTagArrayCheck
+class CacheTagArray : private Log
 {
-    private:
-        /* The set (line) of the cache. */
-        struct CacheTag
-        {
-            Addr line = 0; // data
-            bool is_valid = false; // validity
-        };
-
-        const uint32 num_sets;
-
-        std::vector<std::vector<CacheTag>> array; // array of tags
-        LRUInfo lru; // LRU algorithm module
-
     public:
-        CacheTagArray( uint32 size_in_bytes,
+        CacheTagArray( uint32 size_in_bytes, 
                        uint32 ways,
-                       uint32 line_size = 4,
-                       uint32 addr_size_in_bits = 32);
+                       uint32 size_of_line = 4,
+                       uint32 addr_size_in_bits = 32)
+            : Log( false)
+            , number_of_sets( check_arguments( size_in_bytes, 
```

Comment:
Could you please explain the reason why you removed CacheTagArrayCheck class?

Reply:
Why do we need the separate class just to check arguments?

Reply:
Why not? C++ guarantees that ctor of CacheTagArrayCheck class is executed before anything else, so if arguments are ill-formed, we won't go inside the function and get any undefined behavior. Additionally, if an alternative implementation of CacheTagArray existed (some unusual cache), programmer would inherit from CacheTagArrayCheck and explicitly get all the checks. Instead, your function "check_arguments" may be easily dropped by accident, plus it has very unclear return value (I would expect void or Boolean `true` if everything correct at least).